src/resolver.ts

```diff
@@ -1006,6 +1009,21 @@ export class Resolver extends DiagnosticEmitter {
     ctxType: Type = Type.auto,
     /** How to proceed with eventual diagnostics. */
     reportMode: ReportMode = ReportMode.REPORT
+  ): Type | null {
+    if (this.resolvingExpressions.has(node)) return null;
+
+    this.resolvingExpressions.add(node);
+    const resolved = this.unsafeResolveExpression(node, ctxFlow, ctxType, reportMode);
+    this.resolvingExpressions.delete(node);
+    return resolved;
+  }
+
+  /** Resolves an expression to its static type. (may cause stack overflow) */
+  private unsafeResolveExpression(
```

Comment:
Perhaps...?
```suggestion
  private doResolveExpression(
```
I think the Tokenizer uses a similar `unsafe` prefix for a method, but it's not *technically* unsafe.